src/AI/WallyAI.java

```diff
@@ -0,0 +1,788 @@
+  }
+
+  private boolean isSafe(GameMap gameMap, Map<UnitModel, Map<XYCoord, Double>> threatMap, Unit unit, XYCoord xyc)
+  {
+    Double threat = threatMap.get(myCo.getUnitModel(unit.model.type)).get(xyc);
+    int threshhold = unit.model.hasDirectFireWeapon() ? DIRECT_THREAT_THRESHHOLD : INDIRECT_THREAT_THRESHHOLD;
+    return (null == threat || threshhold > threat);
+  }
+
+  /**
+   * @return whether it's safe or a good place to wall
+   */
+  private boolean canWallHere(GameMap gameMap, Map<UnitModel, Map<XYCoord, Double>> threatMap, Unit unit, XYCoord xyc)
+  {
+    // if we're safe, we're safe
+    if( isSafe(gameMap, threatMap, unit, xyc) )
+      return true;
+    // don't stand on a factory or in danger for no good reason
+    if( gameMap.getEnvironment(xyc).terrainType == TerrainType.FACTORY || !unit.model.hasDirectFireWeapon())
+      return false;
+
+    // TODO: Determine whether the ally actually needs a wall there. Mechs walling for Tanks vs inf is... silly.
+    // if we'd be a nice wall for a worthy ally, we can pretend we're safe there also
+    XYCoord[] adjacentCoords = { new XYCoord(xyc.xCoord + 1, xyc.yCoord), new XYCoord(xyc.xCoord - 1, xyc.yCoord),
+        new XYCoord(xyc.xCoord, xyc.yCoord + 1), new XYCoord(xyc.xCoord, xyc.yCoord - 1) };
+    for( XYCoord coord : adjacentCoords )
+    {
+      Location loc = gameMap.getLocation(coord);
+      if( loc != null )
+      {
+        Unit resident = loc.getResident();
+        if( resident != null && !myCo.isEnemy(resident.CO) 
+            && resident.model.getCost() * resident.getHP() > unit.model.getCost() * unit.getHP() )
+        {
+          return true;
+        }
+      }
+    }
+    return false;
+  }
+
+  /**
+   * Attempts to find a combination of attacks that will create a kill.
+   * Recursive.
+   */
+  private double findAssaultKills(GameMap gameMap, Map<XYCoord, Unit> neededAttacks, Unit target, double pDamage)
+  {
+    // base case; we found a kill
+    if( pDamage >= target.getPreciseHP() )
+    {
+      return pDamage;
+    }
+
+    double damage = pDamage;
+    for( XYCoord xyc : neededAttacks.keySet() )
+    {
+      // Don't try to attack from the same space twice.
+      if( null != neededAttacks.get(xyc) )
+        continue;
+
+      // Attack with the cheapest assault units, if possible.
+      Queue<Unit> assaultQueue = new PriorityQueue<Unit>(11, new AIUtils.UnitCostComparator(true));
+      assaultQueue.addAll(unitQueue);
+      while (!assaultQueue.isEmpty())
+      {
+        Unit unit = assaultQueue.poll();
+        if( !unit.model.hasDirectFireWeapon() || neededAttacks.containsValue(unit) ) // don't try to attack twice with one unit
+          continue;
+
+        int dist = Math.abs(unit.x - target.x) + Math.abs(unit.y - target.y);
+
+        // Figure out how to get here.
+        Path movePath = Utils.findShortestPath(unit, xyc, gameMap);
+
+        if( movePath.getPathLength() > 0 && unit.canAttack(target.model, dist, true) )
```

Comment:
Correct me if I'm wrong, but I don't think this condition can ever pass. Consider:
a) We are specifically looking for mobile attacks in this function; we've already looked at target-adjacent units above on lines 227-241. Movement will be required.
b) `dist` is calculated using the current `unit` and `target` positions, which by a) will not be adjacent. Given that mobile weapons are currently all 1-range, `dist` must therefore always be too far.

I suppose you could pass the check if you had a unit placed outside of direct-fire range, that had both a direct-fire and an indirect-fire weapons, and that somehow didn't use the ranged attack during the siege-assignment phase.
At any rate, I ran a sample game with a breakpoint inside this if, and it never got hit.

Reply:
Huh. I could have sworn I've seen it trigger.

Regardless, the calculation of `dist` is wrong. It should be from `xyc` to `target`. Good catch.